tests/lean/refineFiltersOldMVars.lean

```diff
@@ -0,0 +1,97 @@
+import Lean
+
+/-! These tests ensure that `refine e` only returns goals that were created during elaboration of `e` (and not before).
+
+Including mvars encountered in `e` that were created at any point in time caused trouble in a couple of ways:
+
+* Pre-existing goals were duplicated in the infoview (issue #2495)
+* "Goal tunneling": natural holes (`_`) created far earlier in the term could suddenly resurface after using `refine` on a term that happened to involve them (not documented on the lean4 repo; discovered during testing in Mathlib). A schematic example of this sort of issue:
+```
+def x := {
+  /- `field1` introduces a natural mvar: -/
+  field1 := f _
+  /- the value of `field1` is used in `field2`, and prior to this fix, the goal created above "tunnels" into the infoview via `refine'`: -/
+  field2 := by refine' <term involving field1> -- includes pre-existing natural mvar
+}
+```
+-/
+
+/-! # Issue 2495 -/
+
+/- In the following tests we take advantage of the fact that the unsolved goals error reports the
+tactic state to ensure that goals haven't been duplicated. -/
+
+/-! Refining using an existing goal (in this case, simply unifying it with the main goal) should
+only produce one `case a` in the infoview, not two. -/
+
+example : True := by
+  have : True := ?a
+  refine ?a -- should produce only one `case a`
+
+/-! Only new goals in the refined expression should replace the main goal. Pre-existing side goals should stay at the bottom of the list. -/
+
+inductive Foo where
+| mk (α : Type) : α → Foo
+
+example : Foo := by
+  have h : Type := ?a
+  refine .mk ?a ?b
+  /- `?a` is a pre-existing goal, `?b` is not; `refine` should only return `?b`, and `?a` should
+  remain at the bottom of the goal list. -/
+
+/-! # Goal tunneling -/
+
+/-! Natural mvars created earlier should not be able to tunnel into later uses of `refine`, but should be solved separately. -/
+
+axiom neq3 {n} : n ≠ 3
+
+structure Bar where
+  x : Nat
+  y : x ≠ 3
+
+/- This should not work: -/
+def bar : Bar := {
+  x := _ + 1
+  y := by
+    refine' neq3 -- captures the underscore above
+    exact 0 -- solves it
+}
+
+/-!
+# Issue #2434
+
+Resolving this issue also resolves issue #2434, which identified an inconsistency in *which* pre-existing mvars were being preserved—namely, `refine` would include old pre-existing syntheticOpaque mvars, but not old natural mvars. This would erroneously close goals. Here, we demonstrate that uniformly filtering out old mvars also resolves that issue.
```

Comment:
```suggestion
Resolving this issue also resolves issue #2434, which identified an inconsistency in *which* pre-existing mvars were being preserved --- namely, `refine` would include old pre-existing syntheticOpaque mvars, but not old natural mvars. This would erroneously close goals. Here, we demonstrate that uniformly filtering out old mvars also resolves that issue.
```

Reply:
(I see you used a unicode em-dash, but in practice it's unreadable. :-)